### ISSUE TITLE
Exploring Semantic Spaces - Jurafsky & Martin 2015

### ISSUE THREAD

**Issue by jamesallenevans** [Computational-Content-Analysis-2020/Readings-Responses#36]
Jurafsky, Daniel and James H. Martin. 2015. _Speech and Language Processing_. Chapters 6 & 7 ([“Vector Semantics and Embeddings”](https://web.stanford.edu/~jurafsky/slp3/6.pdf), [“Neural Networks and Neural Language Models”](https://web.stanford.edu/~jurafsky/slp3/7.pdf))

**Comment by lkcao** [comment]
In chapter 6, figure 6.14 shows how to trace the semantic change of words in historical context. "The grey contexts words are trained using most recent time-point embeddings while earlier points are trained from earlier historical embedding spaces." My question is, how to project a word from historical space (for example, 1900s "gay") in the 1990s space? Could you explain the mathematical process?

**Comment by heathercchen** [comment]
I am confused about the hierarchical representation of word embedding shown in section 6.9 (on p.22). 
![image](https://user-images.githubusercontent.com/54857489/74906263-0a2a1d80-5376-11ea-9bfe-0af2c9128973.png)
The origin node of this tree graph is left blank here, what does that imply? As we can see from the graph, "dog" and "cat" are two child nodes of the same father node, which implies they are similar to each other in word embeddings. But the word "puppy" is one node away from the nodes mentioned above, does it mean that the word "puppy" is somewhat related to "dog" but has farther distance than "cat"?

**Comment by deblnia** [comment]
I thought I understood skip-gram models before this, but I was wrong! Could we go into some more detail about the target and the context embeddings and why they need positive/negative examples respectively?

**Comment by katykoenig** [comment]
Chapter 7 states that each neural unit takes the it's input and produces an output via an activation function. The authors further note that the ReLU and tanh activation functions have different properties that make them useful for different language applications. I am wondering if there are other activations we should explore for our project, i.e. are also relevant to language processing, and if there are some hard and fast rules to picking an activation function (or activation functions if many hidden layers) to save time via avoiding much trial-and-error?

**Comment by zikaili** [comment]
I'm wondering when we should conduct debiasing when using word embedding models and how we should approach it. It seems the reproduction of human bias is particularly consequential for the use of embedding models in prediction tasks like the screening of resumes or predicting crime risk in the aid of bail decisions. But how can we establish the appropriate baseline for debiasing to avoid over-correction?

**Comment by laurenjli** [comment]
Due to the many layers in a neural network, the drivers behind a given result or prediction can often be obscure. What are some ways to make neural networks more interpretable?

**Comment by HaoxuanXu** [comment]
It would be interesting to know the runtime of tfidf model compared with word2vec model. Is word2vec more efficient than tfidf due to having fewer sparse matrix?

**Comment by bjcliang-uchi** [comment]
It seems that it is possible to extract the embedding information of the neural network models to somehow summarize the meaning of words. Can we apply this idea to topic modeling as well?

**Comment by di-Tong** [comment]
Chapter 6 mentions that some times researchers throw away context matrix, some times they combine the target matrix and content matrix either in sum or in concatenated form. I wonder if the two matrices differ significantly from each other. What are the implications of choosing one form of embedding matrix for further analyses?

Besides, for Figure 6.14, I wonder if we can directly compare word vectors from different embedding spaces, since the random initialization for each word embedding model would probably affect the resulting vector values?

**Comment by ccsuehara** [comment]
I was wondering if the hidden layers of the neural networks have any interpretability in the context of neural language models.

**Comment by sunying2018** [comment]
I have a question in the backward pass. In this backward updating process, should we use the original weights of the previous layers or the most recent weights (new gradients)? I am wondering which is more conventional way in neural language models?  Since in some material, they only use the original weights for simplicity such as the example in https://mattmazur.com/2015/03/17/a-step-by-step-backpropagation-example/   and there are examples as well iteratively use the most recent value instead of the value of last step.

**Comment by sanittawan** [comment]
For a neural network model covered in chapter 7, I am wondering if it is possible to train a classifier using text as a feature along with other features like age, gender, etc. If so, what is the best way to do it? In what form should the text data be?

**Comment by chun-hu** [comment]
The SGNS embedding algorithm is new to me! I'm wondering if we can talk more about how it allows incremental initialization during learning, which seems to be a very useful function.

**Comment by rachel-ker** [comment]
Given the implicit associations and bias captured in the embeddings via text and the difficulty of debiasing embeddings, would that  suggest that word embeddings are more appropriate to describe the current social context and semantic relations (i.e. situations where the bias is part of the measure)? And not so appropriate where we want a more objective/predictive analysis? Or would there be ways to measure the bias and correct it later?

**Comment by luxin-tian** [comment]
@heathercchen raises a good question regarding the visualization of word2vec. Based on my understanding, this graph is pretty much like hierarchical clustering. We might interpret it as, starting from each word as their own vector, being separated in this hierarchal graph, any two words (vectors) that have not been linked together yet and have the highest similarity get to be linked in the next hierarchy and repeat this step again and again. Overall, it is a visual representation of similarities between any two word-vectors. Therefore, for the specific corpus that gives this graph in the textbook, "cat" is indeed more similar to "dog" than "puppy" is. 

I wonder is it the correct way of interpreting this graph?

**Comment by jsmono** [comment]
I am curious if this method can be applied to nouns composed of two words. This problem has occurred many times during the past few exercises when I wanted to treat names like "hong kong" as one word but the system chose to view it as two and conduct separate analyses on each. Is it possible to avoid such situations when applying words embedding?

**Comment by wunicoleshuhui** [comment]
From an empirical perspective, what are the typical cases in which word2vec is preferred over skipgram as the method? What kinds of data work with word2vec better, and what data would work with skipgram better?

**Comment by cindychu** [comment]
For word embedding, in the assignment, we are suggested do not use lemmatized form of word, however in this chapter, it seems that lemma is where contains the semantic meaning of the words, so I am wondering how lemmatization could influence the word embedding process (‘lexical semantics') and why we do not do lemmatization in our practical implementation?

**Comment by skanthan95** [comment]
Looking at how word2vec learns embeddings:

![Screen Shot 2020-02-21 at 9 51 52 AM](https://user-images.githubusercontent.com/56130450/75049490-e529cd80-548f-11ea-817b-28a9ffda421c.png)

How do we deal with esoteric metaphors or idioms that would likely be classified as negative? Would we have to manually sort through the instances that were negatively classified? I'm also thinking of upscale restaurants that might incorporate something (like apricot) as an ingredient in unorthodox ways, and how word2vec might misclassify that (e.g. apricot, venison). How would the training process differ in a situation/context like that?

**Comment by arun-131293** [comment]
How do we learn embeddings that learn different embeddings for different uses of the word using the methodology described here? For instance, how we do get a Vector Space Model where "played" in "Brazil played Argentina" and "He played Abraham Lincoln" have different embeddings?

**Comment by yirouf** [comment]
I'm curious about Jufarsky and Martin's idea about hierarchy of word embedding. It is somehow similar to human's memory model, associative network, where words are grouped in terms of meaning. I wonder how this could account for other languages. In other languages, there are also single word that means different things (or the same character with different pronunciation). How would that be accounted for?

**Comment by rkcatipon** [comment]
It's exciting to learn more about Word2Vec, as it has often been used in computational social science papers. I have a small question, in regards to  (Vector Semantics and Embeddings” -  the authors reference linguistic foundations that places words along three dimensions:

> valence: the pleasantness of the stimulus 
> arousal: the intensity of emotion provoked by the stimulus 
> dominance: the degree of control exerted by the stimulus

But do these dimensions hold for non-English, or non-Western languages?  For example, if we take words that do not have direct translations in English, such as the Portguese-Brazilian term _Saudade_ ,  coined by Portguese settlers longing for their native land and defined loosely as a nostalgia for a future that never was -- where would these sorts of terms fall along the three dimensions placed above?

**Comment by alakira** [comment]
I am curious what pre-processing is best for skip-gram or other methods. Would lemmatizing or filtering certain pos words could improve the results or rather distort it?

**Comment by YanjieZhou** [comment]
I have some questions about the hierarchical representation of word embedding. What are the similarities and differences between this represenation and the hierarchical clustering that we covered?  What does it actually mean to be seperated into different subcategories? How can we interprete the subtle differences between words in the same subcategory and the ones in the neighboring categories?

**Comment by cytwill** [comment]
I noticed that in many of the word embedding methods mentioned above, we need to specify the window in which we choose words for context. So are there any rules of thumb in choosing such windows or it is better for us to do a sensitivity test for multiple values?  Also, even if with word2vec method, there seem to still quite a large number of dimensions in word vectors.  So I wonder if it is possible to incorporate some dimension reduction (like PCA) during the feature engineering of word vectors?

**Comment by ziwnchen** [comment]
So far when we say dimensions of the word embedding space, what we mean is orthogonal, cartesian coordinate. I understand this might be because of the form of "vector", which contains values for each independent dimension. But just a stupid idea, is it possible to build embedding space on a polarized coordinate? For example, like the circle in basic human value [plot](https://en.wikipedia.org/wiki/Theory_of_Basic_Human_Values#/media/File:Theory_of_Basic_Human_Values_Graphic.jpg)

**Comment by ckoerner648** [comment]
Having been taught grammar in school as a subject on which the book is closed, I was surprised to learn that linguists still develop new forms of grammar and that “there is considerable variation from theory to theory” (p. 3). In light of this fuzziness, is it really correct to parse a sentence in the way that a certain expert linguist would do it? Or are we interested in the way in which the average reader understands the text, which may very well mean how they misunderstand the text in the case of complicated non-projective trees in languages like Czech or German?

**Comment by kdaej** [comment]
Skip-gram assumes that all context words are independent. I am curious about what this means with real text data. Is it just a matter of how to determine the unit of the analysis or does it need to be rigorously examined to make sure they are not related?

**Comment by VivianQian19** [comment]
It is clear from the reading and coding assignments how tf-idf model is achieved in English language corpus. I wonder how this model is applied in languages other than English?
